File: BinarySearchTree/validate-binary-search-tree.js
// https://leetcode.com/problems/validate-binary-search-tree/

//top down
// ========




/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
    let res = true
    function traverse(curr, lo = null, hi = null) {
        if(lo || lo == 0) {
            if(curr.val <= lo) {res = false}
        } 
        if(hi || hi == 0) {
            if(curr.val >= hi) {res = false}
        } 
        curr.left && traverse(curr.left, lo, curr.val)
        curr.right && traverse(curr.right, curr.val, hi)
    }
    traverse(root)
    return res
};




// bottom up
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root, lo = null, hi = null) {
    if (!root) return true

    if((lo != null && root.val <= lo) || (hi != null && root.val >= hi)) return false
    const left = isValidBST(root.left, lo, root.val)
    const right = isValidBST(root.right, root.val, hi)

    return left && right
};

File: Microsoft/validate-binary-search-tree.py
# https://leetcode.com/problems/validate-binary-search-tree/submissions/2095076753/?envType=company&envId=microsoft&favoriteSlug=microsoft-thirty-days
# https://leetcode.com/problems/validate-binary-search-tree/description/?envType=company&envId=microsoft&favoriteSlug=microsoft-thirty-days


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def isValid(root, lo, hi):
            if (not root):
                return True
            if (lo!=None and lo >= root.val) or (hi!=None and hi <= root.val):
                return False
            l = isValid(root.left, lo, root.val)
            r = isValid(root.right, root.val, hi)
            return l and r
        return isValid(root, None, None)